### PR TITLE
[FLINK-36368] Do not prematurely merge CommittableManager

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

When a sink contains a shuffle between writer and committer, a committer may receive committables coming from multiple subtasks. So far, we immediately merged then on receival. However, that makes it later impossible to trace whether we received all messages from an upstream task.

It also made rescaling of the committer awkward: during normal processing all committables of a committer have the same subtaskId as the committer. On downscale, these subtaskIds suddenly don't match and need to be replaced, which we solved by merging the SubtaskCommittableManagers.

This commit decouples the collection of committables from changing the subtaskId for emission. Committables retain the upstream subtask id in the CommittableCollection, which survives serialization and deserialization. Only upon emission, we substitute the subtask id with the one of the emitting committer.

This is, in particular, useful for a global committer, where all subtasks are collected. As a side fix, the new serialization also contains the numberOfSubtasks such that different checkpoints may have different degree of parallelism.

The old approach probably has edge cases where scaling a UC would result in stalled pipelines because certain metadata doesn't match. This would not affect pipelines which  chain Writer/Committer (no channel state), Writer and Committer have same DOP (results in a Forward channel, which doesn't use UC for exactly these reasons), and a non-keyed shuffles (because they don't provide any guarantees). Since a keyed shuffle must use the subtask id of the committables, the new approach should be safe. However, since we disabled UC entirely for sinks to adhere to the contract of notifyCheckpointComplete, this shouldn't matter going forward. It's still important to consider these cases though for restoring from Flink 1 checkpoints.

## Brief change log

* Do not prematurely merge CommittableManager 

## Verifying this change

Covered by the following existing test suites that I partially adjusted
* flink-runtime/src/test/java/org/apache/flink/streaming/runtime/operators/sink/committables
* flink-runtime/src/test/java/org/apache/flink/streaming/api/connector/sink2
* flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/sink

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java [hunk 1]
```diff
@@ -117,7 +126,7 @@ SubtaskCommittableManager<CommT> getSubtaskCommittableManager(int subtaskId) {
     }
 
     @Override
-    public CommittableSummary<CommT> getSummary() {
+    public CommittableSummary<CommT> getSummary(int subtaskId, int numberOfSubtasks) {
         return new CommittableSummary<>(
                 subtaskId,
                 numberOfSubtasks,
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:132]: Something looks strange here. We should rename either the field `numberOfSubtasks` or the method parameter `numberOfSubtasks` to distinguish the number of sub tasks on receival and emission.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java [hunk 2]
```diff
@@ -254,4 +209,31 @@ private CheckpointCommittableManagerImpl<CommT> getCheckpointCommittables(
                 this.checkpointCommittables.get(committable.getCheckpointIdOrEOI());
         return checkNotNull(committables, "Unknown checkpoint for %s", committable);
     }
+
+    /** Removes all metadata about checkpoints of which all committables are fully committed. */
+    public void compact() {
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:214]: Nit: This change looks unrelated to the overall change of this PR. I am okay with removing the side-effect of cleaning the committables on `get`.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollectorSerializer.java [hunk 3]
```diff
@@ -48,18 +48,21 @@ public final class CommittableCollectorSerializer<CommT>
...
+    private final int defaultNumberOfSubtasks;
+
     private final SinkCommitterMetricGroup metricGroup;
 
     public CommittableCollectorSerializer(
             SimpleVersionedSerializer<CommT> committableSerializer,
-            int subtaskId,
-            int numberOfSubtasks,
+            int defaultSubtaskId,
+            int defaultNumberOfSubtasks,
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollectorSerializer.java:61]: I find `default` the wrong word. Can we call it `currentTask[....]`

**Review by fapaul — COMMENTED** [review]
Overall idea looks solid only left some questions regarding the implementation.

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java [hunk 4]
```diff
@@ -160,28 +154,20 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
 
-    private Collection<? extends CheckpointCommittableManager<CommT>> getCommittables(
-            long checkpointId) {
-        final Collection<? extends CheckpointCommittableManager<CommT>> committables =
-                committableCollector.getCheckpointCommittablesUpTo(checkpointId);
-        if (committables == null) {
-            return Collections.emptyList();
-        }
-        return committables;
-    }
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:171]: It's never returning null and it's not marked @Nullable. Tbh not sure why this code was in there.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollectorSerializer.java (hunk 3 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollectorSerializer.java:61]: Did you read the comment on the fields?

```
    /** Default values are used to deserialize from Flink 1 that didn't store the information. */
    private final int defaultSubtaskId;
    /** Default values are used to deserialize from Flink 1 that didn't store the information. */
    private final int defaultNumberOfSubtasks;
```

After the PR we don't need that information anymore and just used for backward compatibility. I actually thought of just setting them to some constant as it shouldn't matter but decided to emulate the old behavior as much as possible.

If you still think `currentSubtaskid` is more appropriate, how about `owningSubtaskId`?

Comment (fapaul) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollectorSerializer.java:61]: I like `owningSubtaskId`

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java (hunk 2 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:214]: Fair point that it's a bit mixed but it was also not big enough to be a commit on its own imho. So I'd leave as-is.

**Review by fapaul — APPROVED** [review]
Lgtm now
